### PR TITLE
[DP] Functional DP for GPT-OSS

### PR DESCRIPTION
**Pull request by wenxindongwork**
# Description

Add functional DP support for GPT-OSS Torchax backend. 

Verified baseline throughput unchanged (5037.82) , DP=2 throughput is 1.54x (7781.92). 

Validated numerical correctness with offline_inference.py

Full details: https://paste.googleplex.com/5240826907197440

# Tests

https://buildkite.com/tpu-commons/tpu-inference-ci/builds/5851

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/vllm/fused_moe.py [hunk 1]
```diff
@@ -167,14 +182,26 @@ def _gmm_all_reduce(lhs, rhs, group_sizes):
...
-        gmm_result = (gmm_result + rhs_bias).astype(gmm_result.dtype)
+
+        def _add_bias(gmm_result_local, rhs_bias_local, group_sizes_global):
+            rhs_bis = jnp.repeat(rhs_bias_local,
+                                 group_sizes_global,
+                                 0,
+                                 total_repeat_length=m // mesh.shape["data"])
+            return (gmm_result_local + rhs_bis).astype(gmm_result_local.dtype)
+
+        gmm_result = shard_map(
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/fused_moe.py:197]: personally, I prefer not using shard_map if possible as it is really prone to numeric error when not properly used. When using `check_rep=False`, unlike other ops, there isn't any safety feature that guarantees that all the numerics of a tensor adheres to a proper SPMD / sharding annotation.

I prefer using it only when it's really necessary (like using kernel).

Please modify this code not use shard map and you can refer to this PR where I replaced existing using of shard_map to a regular jax function: https://github.com/vllm-project/tpu-inference/pull/590

File: tpu_inference/layers/vllm/quantization/common.py [hunk 2]
```diff
@@ -61,7 +61,12 @@ def __init__(self, vllm_config: VllmConfig, mesh: Mesh, layer: LinearBase):
                 " bad performance.", type(layer))
 
         self.bias_sharding = P(self.weight_sharding[0])
-        self.n_shards = self.mesh.shape.get(self.weight_sharding[0], 1)
+        if isinstance(self.weight_sharding[0], tuple):
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/quantization/common.py:64]: is it an delibrate choice to not use `ShardingAxisName`?: https://github.com/vllm-project/tpu-inference/blob/3a9e2d4d34c5205ebd2c2ea913b8d82bc928db76/tpu_inference/layers/common/sharding.py#L50-L57

File: tpu_inference/layers/vllm/fused_moe.py (hunk 1 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:197]: For more context, when using shard map with check_rep=False, when using something like 'out_spec=P(None)', it only annotates the tensor as having that sharding but shard map does not introduce any collective to ensure it.

Meaning, it is possible that the output tenor numeric is not replicated along multiple devices and all  devices have different numeric because shard_map does not provide any guarantees - which makes it really painful to debug when there's a numeric issue.

Comment (wenxindongwork) [reply tpu_inference/layers/vllm/fused_moe.py:197]: I tried to not use a shard_map but did not figure out a way that doesn't involve a for loop. 

The complexity here is that `jnp.repeat(bias, group_size, 0)` expects bias and group_size to share the same size on dimension 0, but group_size[0]= `DP*num_experts` whereas `bias[0] = num_experts`.

File: tpu_inference/layers/vllm/quantization/common.py (hunk 2 above)
Comment (wenxindongwork) [reply tpu_inference/layers/vllm/quantization/common.py:64]: I wanted to make this PR minimal and debuggable. I plan to change to ShardingAxisName when I add attn_dp support, does that sounds like a good plan to you?

Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/common.py:64]: sgtm!

File: tpu_inference/layers/vllm/fused_moe.py (hunk 1 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:197]: if that's the case, can you do something like this?

```
# convert (experts, model_dim) to (experts * dp_size, model_dim)
bias = jnp.repeat(bias, dp_size, 0) 

# (optional. may or may not needed) match bias's sharding with group_size's sharding
bias = jax.lax.with_sharding_constraint(bias, P("data", "model"))

# Now the bias.shape[0] and group_size.shape[0] matches
rhs_bias = jnp.repeat(bias, group_size, 0)
```

**Review by kyuyeunk — COMMENTED** [review]
> Please add torchax dp related unit tests.

Also, please address this comment.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 1 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:197]: Note that if complier optimization works correctly, the first 2 code (jnp.repeat & sharding constraint) will be a no-op. because the data is already present in each dp rank & we are just telling it to treat them differently starting from now.

Comment (wenxindongwork) [reply tpu_inference/layers/vllm/fused_moe.py:197]: Thanks Kyuyeun for this suggestio, For correctness, I have to use `jnp.tile` instead of `jnp.repeat`. however, I am noticing performance drop (7575.50 vs 7781.92) if I do this instead of shard_map. Maybe due to `jnp.tile` not being a no-op? 

```
        rhs_bias = jnp.tile(rhs_bias, (mesh.shape["data"], 1)) 
        # adding the sharding constraint does not make a difference 
        rhs_bias = jnp.repeat(rhs_bias, group_sizes, 0, total_repeat_length=m)
        gmm_result = (gmm_result + rhs_bias).astype(gmm_result.dtype)
```

**Comment by wenxindongwork** [comment]
added e2e model parallelism test for Llama3.1 1b for torchax.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 1 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:197]: Hmm in-theory it should be no-op.

because bias is already replicated along TPUs in dp axis, and combining tile/repeat with sharding constraint just tells TPU to treat them like a separate non-replicated tensor.

I'll do some test locally and get back to you asap.

Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:197]: okay, i've done various things and was honestly surprised that tile/repeat method didn't work??

so let's leave it as using shard_map for now and figure out a if we can do something differently later.

but can you remove `check_rep=False` arguments? Because that's what really scares be because it means compiler will not do any checks to ensure numerics are correct.

Comment (wenxindongwork) [reply tpu_inference/layers/vllm/fused_moe.py:197]: done!

**Comment by kyuyeunk** [comment]
i think it all looks good. but can you update the branch, run the ci & confirmed that it all passes?
